### DISCUSSION TITLE
How to pause consuming for a while?

### DISCUSSION THREAD

**Discussion by AhmedAbouelkher** [nats-io/nats-server#3594] [Q&A] [answered]
Hi,

I have a very simple use case, I want to pause a consumer for a while (almost 5 hrs) to do some maintenance and fixes and at the same time to prevent any message loss.

I'm using Jetstream with pre-created durable consumers.

**Comment by bruth** [comment, marked as answer]
Hi @AhmedAbouelkher, the mental model for a consumer is that its a service embedded in the NATS server that keeps track of messages that are consumed (by subscribers). To _pause_ consumption, you can simply unsubscribe/stop the processes that are doing the actual message processing. Once you start them again, consumption will resume.

**Reply by bruth** [reply]
Just to further reinforce with code.. creating the consumer simply sets up tracking of consumption state on the server...

```go
js.AddConsumer("my-stream", &nats.ConsumerConfig{
  Name: "my-consumer",
  AckPolicy: nats.AckExplicitPolicy,
  //etc...
})
```
.. but until a subscription is bound, no message processing occurs.

```go
sub, _ := js.PullSubscribe("", "", nats.Bind("my-stream", "my-consumer"))
msgs, _ := sub.Fetch(5)
```

If at any point your subscribers are not active, no consumption is occurring. When they come back online, they will resume automatically.

**Comment by tanguc** [comment]
You can now pause a consumer with (JS)
``` javascript
// jsm for jetstreamManager
await jsm.consumers.pause(
    MyStreamName,
    MyConsumerName,
    new Date(Date.now() + 24 * 60 * 60) // optional
  );
```

here are whereabouts for each supported languages/sdk: https://github.com/nats-io/nats-architecture-and-design/issues/266

**Comment by Jarema** [comment]
@tanguc is right - we added Consumer Pause, but you need to use preview/nightly server release, as it will be part of 2.11.
